includes/functions.php

```diff
@@ -624,34 +595,97 @@ function tsml_geocode_google($address, $tsml_map_key)
 		]);
 	}
 
+	//get provider name
+	$provider = $tsml_geocoding_methods[$tsml_geocoding_method];
+
 	//start list of options for geocoding request
-	$options = [
-		'key' => $tsml_map_key,
+	$geocode_parameters = [
 		'address' => $address,
 		'language' => $tsml_language,
+		'bounds' => $tsml_bounds
+			? $tsml_bounds['south'] . ',' . $tsml_bounds['west'] . '|' . $tsml_bounds['north'] . ',' . $tsml_bounds['east']
+			: null,
 	];
 
-	//bias the viewport if we know the bounds
-	if ($tsml_bounds) {
-		$options['bounds'] = $tsml_bounds['south'] . ',' . $tsml_bounds['west'] . '|' . $tsml_bounds['north'] . ',' . $tsml_bounds['east'];
-	}
+	//tsml_geocoding_method can be "api_gateway", "legacy", "google_key", or "c4r_geocoding"
+	if ($tsml_geocoding_method === 'api_gateway') {
+		curl_setopt($tsml_curl_handle, CURLOPT_URL, 'https://api-gateway.apps.itstechnical.net/api/geocode?' . http_build_query($geocode_parameters));
+	} elseif ($tsml_geocoding_method === 'c4r_geocoding') {
+
+		if (!$registration = get_option('tsml_geocoding_registration')) {
+
+			//create registration request
+			$registration_parameters = [
+				"organizationName" => get_bloginfo('name'),
+				"websiteURL" => get_bloginfo('url'),
+				"metadata" => [
+					[
+						"keyName" => "appName",
+						"value" => "TSML",
+					], [
+						"keyName" => "appVersion",
+						"value" => TSML_VERSION,
+					], [
+						"keyName" => "wordpressVersion",
+						"value" => get_bloginfo('version'),
+					],
+				]
+			];
+
+			//gateway only supports certain program values
```

Comment:
Technically, these are not currently checked on the backend at run-time, and I just slapped a few in to help define them for developers. But, it would be good to have a run-time check here in which case, these are the ones that should be added:

Program types to add:
ACA
CMA
CoDA
CEA-HOW
DA
DAA
GA
HA
OA
PAL
RCA
RD
RR
SAA
SCA
SLAA
VA

And maybe SG?

Reply:
ok done 👍